etna/core/middleware.py

```diff
@@ -0,0 +1,43 @@
+from django.conf import settings
+from django.template.response import SimpleTemplateResponse
+
+from etna.ciim.utils import get_date_for_retry_after_header
+
+
+def get_client_ip(request) -> str:
+    x_forwarded_for = request.META.get("HTTP_X_FORWARDED_FOR")
+    if x_forwarded_for:
+        ip = x_forwarded_for.split(",")[0]
+    else:
+        ip = request.META.get("REMOTE_ADDR")
+    return ip
+
+
+class MaintenanceModeMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        """
+        Renders for a 503 if MAINTENANCE_MODE is set.
+        Maintenanc Mode is overridden with MAINTENENCE_MODE_ALLOW_IPS
+        """
+        if settings.MAINTENANCE_MODE:
+            # check override maintenance mode
+            if get_client_ip(request) not in settings.MAINTENENCE_MODE_ALLOW_IPS:
+                kwargs = {"template": "503.html", "status": 503}
+                if settings.MAINTENENCE_MODE_ENDS:
+                    kwargs.update(
+                        {
+                            "headers": {
+                                "Retry-After": get_date_for_retry_after_header(
+                                    settings.MAINTENENCE_MODE_ENDS
+                                )
+                            }
+                        }
+                    )
```

Comment:
```suggestion
                try:
                    end_datetime = datetime.datetime.fromisoformat(settings.MAINTENENCE_MODE_ENDS)
                except ValueError:
                    end_datetime = None
                if end_datetime:
                    kwargs.update(
                        {
                            "headers": {
                                "Retry-After": end_datetime.strftime(HTTP_HEADER_FORMAT)
                            }
                        }
                    )
```

Reply:
Would it be fine to retain the `if settings.MAINTENENCE_MODE_ENDS` as I think it may be unnecessary overhead for every request to evaluate `end_datetime`

Reply:
Yes, that'd be fine. In that case, I would probably suggest:

```
                end_datetime = None
                if settings.MAINTENENCE_MODE_ENDS:
                    try:
                        end_datetime = datetime.datetime.fromisoformat(settings.MAINTENENCE_MODE_ENDS)
                    except ValueError:
                        end_datetime = None
                if end_datetime:
                    kwargs["headers"] = {
                        "Retry-After": end_datetime.strftime(HTTP_HEADER_FORMAT)
                    }
```